pkg/tsdb/loki/loki.go

```diff
@@ -0,0 +1,169 @@
+	"github.com/prometheus/common/model"
+)
+
+type LokiExecutor struct{}
+
+func NewLokiExecutor(dsInfo *models.DataSource) (tsdb.TsdbQueryEndpoint, error) {
+	return &LokiExecutor{}, nil
+}
+
+var (
+	plog               log.Logger
+	legendFormat       *regexp.Regexp
+	intervalCalculator tsdb.IntervalCalculator
+)
+
+func init() {
+	plog = log.New("tsdb.loki")
+	tsdb.RegisterTsdbQueryEndpoint("loki", NewLokiExecutor)
+	legendFormat = regexp.MustCompile(`\{\{\s*(.+?)\s*\}\}`)
+	intervalCalculator = tsdb.NewIntervalCalculator(&tsdb.IntervalOptions{MinInterval: time.Second * 1})
+}
+
+func (e *LokiExecutor) Query(ctx context.Context, dsInfo *models.DataSource, tsdbQuery *tsdb.TsdbQuery) (*tsdb.Response, error) {
+	result := &tsdb.Response{
+		Results: map[string]*tsdb.QueryResult{},
+	}
+
+	client := &client.DefaultClient{
+		Address:  dsInfo.Url,
+		Username: dsInfo.BasicAuthUser,
+		Password: dsInfo.DecryptedBasicAuthPassword(),
+	}
+
+	queries, err := parseQuery(dsInfo, tsdbQuery.Queries, tsdbQuery)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, query := range queries {
+
+		plog.Debug("Sending query", "start", query.Start, "end", query.End, "step", query.Step, "query", query.Expr)
+		span, _ := opentracing.StartSpanFromContext(ctx, "alerting.loki")
+		span.SetTag("expr", query.Expr)
+		span.SetTag("start_unixnano", query.Start.UnixNano())
+		span.SetTag("stop_unixnano", query.End.UnixNano())
+		defer span.Finish()
+
+		//currently hard coded as not used - applies to log queries
+		limit := 1000
+		//currently hard coded as not used - applies to queries which produce a stream response
+		interval := time.Second * 1
+
+		value, err := client.QueryRange(query.Expr, limit, query.Start, query.End, logproto.BACKWARD, query.Step, interval, false)
+
+		if err != nil {
+			return nil, err
+		}
+
+		queryResult, err := parseResponse(value, query)
+		if err != nil {
+			return nil, err
+		}
+		result.Results[query.RefId] = queryResult
+	}
+
+	return result, nil
+}
+
+func formatLegend(metric model.Metric, query *LokiQuery) string {
+	if query.LegendFormat == "" {
+		return metric.String()
+	}
+
+	result := legendFormat.ReplaceAllFunc([]byte(query.LegendFormat), func(in []byte) []byte {
+		labelName := strings.Replace(string(in), "{{", "", 1)
+		labelName = strings.Replace(labelName, "}}", "", 1)
+		labelName = strings.TrimSpace(labelName)
+		if val, exists := metric[model.LabelName(labelName)]; exists {
+			return []byte(val)
+		}
+		return []byte{}
+	})
+
+	return string(result)
+}
+
+func parseQuery(dsInfo *models.DataSource, queries []*tsdb.Query, queryContext *tsdb.TsdbQuery) ([]*LokiQuery, error) {
+	qs := []*LokiQuery{}
+	for _, queryModel := range queries {
+		expr, err := queryModel.Model.Get("expr").String()
+		if err != nil {
+			return nil, err
+		}
+
+		format := queryModel.Model.Get("legendFormat").MustString("")
+
+		start, err := queryContext.TimeRange.ParseFrom()
+		if err != nil {
+			return nil, err
```

Comment:
We could probably wrap these parsing error to be more descriptive. Like `fmt.Errorf("failed to parse From: %v", err)`.